src/FormattedMessage.js

```diff
@@ -0,0 +1,39 @@
+const format = (ctx, values) => ctx.parent.$formatMessage(ctx.props, values)
+const placeholder = (name) => `@\0@\0${name}\0@\0@`
+const placeholderRegex = /@\0@\0(.*?)\0@\0@/g
+
+export default {
+    functional: true,
+
+    props: {
+        id: {type: String, required: true},
+        defaultMessage: String,
+        values: Object,
+        tagName: {type: String, default: 'span'}
+    },
+
+    render (h, ctx) {
+        const slots = ctx.slots()
+        const slotNames = Object.keys(slots)
+        if (slotNames.length === 0) {
+            return h(ctx.props.tagName, format(ctx, ctx.props.values))
+        }
+
+        const values = Object.assign({}, ctx.props.values)
```

Comment:
The local component scope when these are used is actually the parent's scope (because everything is done inline in the parent template). Can't think of a better way to do this now, so we'll stay parallel with React Intl implementation.